nipype/utils/profiler.py

```diff
@@ -0,0 +1,333 @@
+# -*- coding: utf-8 -*-
+# @Author: oesteban
+# @Date:   2017-09-21 15:50:37
+# @Last Modified by:   oesteban
+# @Last Modified time: 2017-09-28 13:11:03
+"""
+Utilities to keep track of performance
+"""
+from __future__ import print_function, division, unicode_literals, absolute_import
+
+import threading
+from time import time
+try:
+    import psutil
+except ImportError as exc:
+    psutil = None
+
+from .. import config, logging
+from .misc import str2bool
+from builtins import open, range
+
+proflogger = logging.getLogger('utils')
+
+resource_monitor = str2bool(config.get('execution', 'resource_monitor', 'false'))
+if resource_monitor and psutil is None:
+    proflogger.warn('Switching "resource_monitor" off: the option was on, but the '
+                    'necessary package "psutil" could not be imported.')
+    resource_monitor = False
+
+# Init variables
+_MB = 1024.0**2
+
+
+class ResourceMonitor(threading.Thread):
+    def __init__(self, pid, freq=5, fname=None):
+        if freq <= 0:
+            raise RuntimeError('Frequency (%0.2fs) cannot be lower than zero' % freq)
+
+        if fname is None:
+            fname = '.nipype.prof'
+
+        self._pid = pid
+        self._fname = fname
+        self._freq = freq
+
+        self._log = open(self._fname, 'w')
```

Comment:
what if we put this in memory or added a timestamp to _fname? wouldn't that take care of nested profilers? i would really try and avoid `resource_profile=False` flag if possible.

Reply:
I see it independently. I think the `resource_profile=False` flag is a necessary tool that ensures you an interface will not trigger the profiler at all, regardless nipype's config.

Then we have the issue of accidentally having two monitors logging to the same file. We have identified the case of nested interfaces because it quickly showed up, but there might be other corner cases in complex workflows that we haven't seen yet. For this reason, I agree we need to work more on this.

@satra, is there a way of writing this to memory, as you say, but making sure that everything that was traced gets finally written to a file no matter what?

Reply:
Also, with writing to memory we will miss a new feature that would allow you keep track of a workflow being run by looking at these profiler files.

Reply:
how about doing: `self._fname = fname + pid + time()`

wouldn't that allow nested profilers?